WithWarnings uses EnsoHashMap to speed things up

Akirathan opened the pull request:
Closes #8682

### Pull Request Description

Majority of warnings handling is now done via newly introduced nodes. Moreover, the underlying representation of warnings storage in `WithWarnings` was changed from `Warning[]` to `EnsoHashMap`.

### Important Notes

- Remove `ArrayRope`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Ensure no regressions in engine and stdlib benchmarks

Commit:
Refactor if instanceof to pattern-match variant

Commit:
WithWarnings uses EnsoHashMap to store warnings

Commit:
MethodProcessor creates HashMapInsertNode in every builtin method

Commit:
Annotations processor does not throw error, but uses processing env messenger instead

Commit:
Make some cached nodes in Warning Shared

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeCallableNode.java:
```diff
@@ -88,7 +92,7 @@ Object invokeWithWarnings(
 
       Warning[] extracted = warnings.getWarnings(warning, null, false);
-      return WithWarnings.wrap(EnsoContext.get(this), result, extracted);
+      return WithWarnings.wrap(result, EnsoContext.get(this), insertNode, interop, extracted);
```
It might be better to create `WithWarnings.WrapNode` and hold the `insertNode` inside of that `Node`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeConversionNode.java:
```diff
@@ -173,7 +176,7 @@ Object doWarning(
             isTail,
             thatArgumentPosition);
-    return WithWarnings.appendTo(EnsoContext.get(this), result, warnings);
+    return WithWarnings.appendTo(result, EnsoContext.get(this), insertNode, interop, warnings);
```
Dtto - basicaly we should wrap the operations `WithWarnings` into Node wrappers and either use their `getUncached()` or `create` version.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Array.java:
```diff
@@ -198,10 +203,13 @@ Warning[] getWarnings(
   }
 
   @CompilerDirectives.TruffleBoundary
```
This method is still behind `@TruffleBoundary`...

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:
```diff
@@ -74,17 +78,30 @@ public EnsoObject getReassignments() {
       autoRegister = false)
   @Builtin.Specialize
   public static WithWarnings attach(
```
The builtin method shall probably be separated from a method (or rather a Node) we use internally in the Engine to handle `attach`.

Akirathan replied:
Handled by [AppendWarningNode](https://github.com/enso-org/enso/blob/02bf7dff8847f4ee6042f5083dc092fd2b101f9a/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java)

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:
```diff
@@ -74,17 +78,30 @@ public EnsoObject getReassignments() {
   @Builtin.Specialize(fallback = true)
-  public static WithWarnings attach(EnsoContext ctx, Object value, Object warning, Object origin) {
-    return WithWarnings.wrap(ctx, value, new Warning(warning, origin, ctx.nextSequenceId()));
+  public static WithWarnings attach(
```
These methods should be deleted and replaced with approriate `Node` types doing the same and specializing internally.

Akirathan replied:
Replaced with [AppendWarningNode](https://github.com/enso-org/enso/blob/02bf7dff8847f4ee6042f5083dc092fd2b101f9a/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java)

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/Warning.java:
```diff
@@ -134,18 +160,34 @@ public static boolean limitReached(WithWarnings value, WarningsLibrary warnings)
+  }
+
+  public static Warning[] fromSetToArray(EnsoHashMap set, InteropLibrary interop) {
+    var vec = set.getCachedVectorRepresentation();
```
We don't need `InteropLibrary` to access internals of `EnsoHashMap`. This code can certainly be simplified.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java:
```diff
@@ -49,7 +53,7 @@
 public final class WithWarnings implements EnsoObject {
   final Object value;
-  private final EconomicSet<Warning> warnings;
+  private final EnsoHashMap warnings;
```
This is correct change!

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java:
```diff
@@ -175,50 +250,138 @@ public Warning[] getWarningsArray(WarningsLibrary warningsLibrary, boolean shoul
+            location,
+            shouldWrap,
+            null,
+            HashMapInsertNodeGen.getUncached(),
```
First question I had when looking at this method: Is `ArrayRope` ready for PE? Answer, probably it is - it is used for concatenating `Text`, isn't it?

Next thought: We are using `getUncached()` anyway. Isn't it slowing things down?

Akirathan replied:
`ArrayRope` was completely removed in 1266cbd2e4fee5b47193eb1086b80ba8cf451ff2

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java:
```diff
@@ -175,50 +250,138 @@ public Warning[] getWarningsArray(WarningsLibrary warningsLibrary, boolean shoul
+   * @param warnings
+   * @return
+   */
+  public static WithWarnings appendTo(Object target, ArrayRope<Warning> warnings) {
```
Replace this new method with `AttachToNode`, if possible.

Akirathan replied:
All `appendTo` and similar methods that append warnings are moved inside a new dedicated [AppendWarningNode](https://github.com/enso-org/enso/blob/02bf7dff8847f4ee6042f5083dc092fd2b101f9a/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java)

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java:
```diff
@@ -324,40 +489,46 @@ public int hashCode(Object o) {
+      return initial;
+    }
+    var set = EnsoHashMap.empty();
+    var initialVec = initial.getCachedVectorRepresentation();
```
We should be able to manipulate with `EnsoHashMap` without using `InteropLibrary`.

JaroslavTulach commented on lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/MethodProcessor.java:
```diff
@@ -282,27 +289,38 @@ private void generateCode(MethodDefinition methodDefinition) throws IOException
         out.println("      @Override");
         out.println("      public Object call(VirtualFrame frame, Object[] args) {");
-        out.println("        return handleExecute(frame, extra, body, args);");
+        out.println("        return handleExecute(frame, extra, body, insertNode, interop, args);");
```
Interesting. I can't say I like this at all! I don't want new arguments to `handleExecute`, if possible... can we keep `handleExecute` untouched and just wrap all calls to `handleExecute` with proper _warnings handling magic_?

Akirathan replied:
That sounds very difficult. `handleExecutes` strips warnings from almost all the arguments, and then reassembles these warnings into the result. Doing this outside of `handleExecute` would mean that we would need to process arguments also outside `handleExecute`.

Why don't you like additional arguments to `handleExecute`?

JaroslavTulach commented on lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/MethodProcessor.java:
```diff
@@ -347,7 +365,9 @@ private void generateCode(MethodDefinition methodDefinition) throws IOException
         out.println("      EnsoContext ctx = EnsoContext.get(bodyNode);");
-        out.println("      return WithWarnings.appendTo(ctx, result, gatheredWarnings);");
+        out.println(
+            "      return WithWarnings.appendTo(result, ctx, insertNode, interop,"
```
Maybe this could be done after the return from `handleExecute`?

Akirathan replied:
Not easily, see the discussion in https://github.com/enso-org/enso/pull/10555#discussion_r1704269605

JaroslavTulach reviewed (commented):
Pavel, you [wrote](https://github.com/enso-org/enso/issues/8682#issuecomment-2233964752):

> the performance is much worse so far, will investigate closely tomorrow

I think the approach you are using is correct. I cannot say why it is slow.

I'd like to advocate usage of specialized `Node` classes to manipulate `WithWarnings`. Ideally `WithWarnings` is an _opaque class_ without any methods that can only be manipulated by associated nodes.

Commit:
Refactor WarningsTest to use ContextUtils

Commit:
Move all warning-related classes to new package

Commit:
WarningsLibrary is part of the type system

Commit:
Use new HasWarningsNode instead of WarningsLibrary.hasWarnings

Commit:
Remove WarningsLibrary.hasWarnings message

Commit:
Fix infinite recursion in ArraySlice.readArrayElement

Commit:
GenerateUncached variants for ArrayLike nodes

Commit:
Revert "Fix infinite recursion in ArraySlice.readArrayElement"

This reverts commit a0dc44ea41b199a2df52a65d2e10a05e1e5a7659.

Commit:
Revert "Remove WarningsLibrary.hasWarnings message"

This reverts commit c4ab6b3ddbf3f5620014ea64a6762a78201aba8d.

Commit:
Revert "Use new HasWarningsNode instead of WarningsLibrary.hasWarnings"

This reverts commit a5dba6f1f0eccb5d54e0edd3099279c50df47b0d.

Commit:
Replace warning.attach by dedicated AppendWarningNode

Removed methods:
- WithWarnings.wrap
- WithWarnings.appendTo
- WithWarnings.append

Commit:
Remove unused WithWarnings constructors

Commit:
Remove unused methods

Commit:
Array.collectAllWarnings is not behind TruffleBoundary

Commit:
Make sure WarningsLibrary is correctly renamed in builtin method processor

Commit:
Add specializations for a single warning to AppendWarningNode

Commit:
Warning.attach handles generic Object as value

Commit:
ArrayRope implements InteropLibrary

Commit:
Warning.get_all is implemented as dedicated node

Commit:
fmt

Commit:
Fix typo in BuiltinMethod annotation

Commit:
Add specialization for Warning[] array to AppendWarningNode

Commit:
Warning.set is implemented as a dedicated Node

Commit:
Do not use uncached nodes at some places

Commit:
Merge branch 'refs/heads/develop' into wip/akirathan/8682-with-warnings-hashmap

Commit:
fmt

Commit:
Warnings hash map have sequenceID as key

Commit:
AppendWarningNode respects warnings limit

Commit:
Cannot insert null as hash map value

Commit:
Add assert not null to method in WithWarnings

Commit:
When wrapping warnings, ensure that the previous warning are sorted

Commit:
Fix specializations in GetAllWarningsNode

Commit:
Fix specializations in SetWarningsNode

Commit:
Remove unused method

Commit:
IsLimitReach is a dedicated node

Commit:
SortVectorNode creates WithWarnings with specific limit

Commit:
fmt

Commit:
profile in SetWarningsNode is shared

Commit:
Fix specializations of AppendWarningNode.

They are now disjunct.

Commit:
Warnings test is executed in context.

Because it executes nodes.

Commit:
Merge branch 'refs/heads/develop' into wip/akirathan/8682-with-warnings-hashmap

# Conflicts:
#	engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java

Commit:
Ensure Array.sort is not on fast-path.

This fixes native image build.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/WarningsTest.java:
```diff
@@ -57,27 +52,28 @@ public static void disposeContext() {
+      var warn2 = Warning.create(ensoContext, "w2", this);
+      var value = 42L;
+
+      var with1 = AppendWarningNodeGen.getUncached().execute(null, value, warn1);
```
You may want to put `getUncached()` into `AppendWarningNode`. That would allow us to _not reference_ the generated class in the code other than `AppendWarningNode` itself.

Akirathan replied:
Done in 82f8506e7ba28723b80f1e5b46dc4644c69aa6af

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java:
```diff
@@ -0,0 +1,159 @@
+   *     It is expected that all the elements in the container are of {@link Warning} class.
+   * @return A wrapped object with warnings
+   */
+  public abstract WithWarnings execute(VirtualFrame frame, Object object, Object warnings);
```
I'd give the method some more specific name. `executeAppend`?

Akirathan replied:
Done in 5003a916b708c2db0b3dfaea3f339df39ccf571c

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/GetAllWarningsNode.java:
```diff
@@ -0,0 +1,73 @@
+    return GetAllWarningsNodeGen.create();
+  }
+
+  public abstract Object execute(@AcceptsWarning Object value, boolean shouldWrap);
```
`executeAllWarnings` maybe?

Akirathan replied:
`BuiltinMethodProcessor` cannot handle methods with different name: it expects there is an `execute` method:
```
[error] /home/pavel/dev/enso/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/GetAllWarningsNode.java:27:1:  error: No execute method found.
[error] public abstract class GetAllWarningsNode extends Node {
[error]                 ^1 error
[warn] frgaal exited with exit code 1
```

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/SetWarningsNode.java:
```diff
@@ -0,0 +1,118 @@
+      Warning[] warnings,
+      @Shared @CachedLibrary(limit = "3") InteropLibrary interop,
+      @Shared @Cached HashMapInsertNode mapInsertNode,
+      @Shared @Cached ConditionProfile isWithWarnsProfile) {
```
There is [InlinedConditionProfile](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/profiles/InlinedConditionProfile.html) now in the Truffle API. It may be more lightweight and the rewrite seems straightforward.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java:
```diff
@@ -74,116 +76,39 @@ public EnsoObject getReassignments() {
+  public static Warning[] fromMapToArray(
+      EnsoHashMap map, ArrayLikeLengthNode lengthNode, ArrayLikeAtNode atNode) {
+    var vec = map.getCachedVectorRepresentation();
+    var vecLen = Math.toIntExact(lengthNode.executeLength(vec));
```
Eh, will we crash the interpreter with `ArithmeticException` when the value doesn't fit into `int`?

Akirathan replied:
900b091fa1ccbb5e6b8a479bb8403e1f9674e5d0 just casts the `long` to `int`. It is very unlikely that there is a long that is greater than Integer.MAX anyway.

JaroslavTulach reviewed (commented):
What are the benchmark results? Not good yet. I see

`Invoke!#Warning.reassign`

and a lot of array creation before it. We need to find a way to put `reassign` to fast path.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/EnsoHashMap.java:
```diff
@@ -66,7 +66,7 @@ EnsoHashMapBuilder getMapBuilder(
   }
 
-  Object getCachedVectorRepresentation() {
+  public Object getCachedVectorRepresentation() {
```
Maybe there was a reason why the method wasn't public! It does not have any reasonable performance. I see `NewArray` allocation coming from this method in the graphs. The usage of `ConditionProfile.getUncached()` probably also doesn't help. The size of the graphs - they are huge!

Why do we want to obtain `Warning[]` as array when it is enough to have them as a snapshot inside of the `EnsoHashMap` anyway?

JaroslavTulach replied:
Please try to get the API of `hash` (and also `vector`) packages where it used to be. The objects shall be opaque. Operations on them should be accessible thru nodes.

Akirathan replied:
All public methods of `EnsoHashMap` and `EnsoHashMapBuilder` refactored to package-protected in these commits:
- 39f7262a7ed6772d542d043be3d9ec63f7ec0353
- 2d7a8b3c571e6545d6ed8cdb065179f897019d70
- 0ecab16a50462fecb3ae1cfae0fcaca0c9f53ac3

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/EnsoHashMap.java:
```diff
@@ -98,7 +98,7 @@ boolean hasHashEntries() {
 
   @ExportMessage
-  int getHashSize() {
+  public int getHashSize() {
```
Why do we need this method `public`. It intentionally wasn't.

Akirathan replied:
Reverted to package protected. See my comment at https://github.com/enso-org/enso/pull/10555#discussion_r1705253874

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeCallableNode.java:
```diff
@@ -88,7 +89,7 @@ Object invokeWithWarnings(
               isTail);
 
       Warning[] extracted = warnings.getWarnings(warning, null, false);
```
Looks like the **most common operation** isn't _append_, but _reassigning_ of warnings from one value to another. We have to **optimize for reassigning**.

If we want to keep decent performance, than we have to stop playing with `Warning[]` arrays. Extracting warnings should be _single instruction operation_ - just **obtain a reference** to `EnsoHashMap`.

The same applies to _appending_ - it should use the **same reference** to `EnsoHashMap` and associate it with the `result`. It should perform _no array manipulation_ at all.

Akirathan replied:
As of 01381e6cc223801e389f37d6bd65ede531475e3d, the solution you propose here is already implemented. Moreover, *reassignments* were already removed altogether - we no longer keep track of the reassignments in the warnings. Resolving this conversation as outdated.

Commit:
Remove reassignments from Warning

Akirathan commented:
Removing reassignments in https://github.com/enso-org/enso/pull/10555/commits/71997339b684cc76d1d1f598233b498e691a95c4 helped the performance: previously score 558 ms/op, now 390 ms/op. I still need to deal with the `EnsoHashMap.getCachedVectorRepresentation()` as that is too expensive. Note that the score of the benchmark on develop is 176 ms/op (and we need to get at least 100x better than that at the end).

Commit:
WarningsLibrary.getWarnings returns reference to EnsoHashMap, not an array.

This should avoid unnecessary new array allocations.

Commit:
Use HashMapInsertAllNode to concat two hash maps.

Avoid materializing entries vector for them.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -520,7 +525,9 @@ Object doPolyglot(
-          accumulatedWarnings = accumulatedWarnings.append(warnings.getWarnings(r, false));
+          EnsoHashMap rWarnsMap = warnings.getWarnings(r, false);
+          Warning[] rWarnsArray = Warning.fromMapToArray(rWarnsMap, lengthNode, atNode);
+          accumulatedWarnings = accumulatedWarnings.append(rWarnsArray);
```
A conversion to `Warning[]` remains. Not sure why? If we can obtain a `EnsoHashMap` for two (or more) values, then we can always _add them together_ as `EnsoHashMap`, can't we?

Akirathan replied:
Removed in 01381e6cc223801e389f37d6bd65ede531475e3d

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/HashMapInsertAllNode.java:
```diff
@@ -0,0 +1,62 @@
+    shortName = "MapInsertAll",
+    description = "Inserts all elements from the given container into a hash map",
+    language = LanguageInfo.ID)
+public abstract class HashMapInsertAllNode extends Node {
```
Yeah, a node to put two `EnsoHashMap` together (even with a limit) is going to be needed!

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java:
```diff
@@ -94,12 +96,58 @@ WithWarnings doMultipleWarningsArray(
+    assert !(object instanceof WithWarnings);
+    int warnsLimit = EnsoContext.get(this).getWarningsLimit();
+    return new WithWarnings(
+        object, warnsLimit, newWarnsMap.getHashSize() >= warnsLimit, newWarnsMap);
```
Can someone instantiate new `WithWarnings` with a different limit than `EnsoContext.get(this).getWarningsLimit()`? If not, that we could hide all this limit manipulation inside of the `WithWarnings` class.

Akirathan replied:
So far, only [SortVectorNode](https://github.com/enso-org/enso/blob/788115f79311068d2f5eb774f93672a4de207178/engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java#L359) uses a different warning limit than the one stored in the `EnsoContext`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java:
```diff
@@ -94,12 +96,58 @@ WithWarnings doMultipleWarningsArray(
+      return withWarnings;
+    }
+    var maxWarns = withWarnings.maxWarnings;
+    var newWarnsEntriesVec = newWarnsMap.getCachedVectorRepresentation(hasCachedVecReprProfile);
```
I expect the `getCachedVectorRepresentation` to be quite slow operation. Given one just iterates thru it, we might rather use _iterator_ like concept.

Akirathan replied:
Using `HashMapInsertAllNode` in c2815e8a5d1064c2bbce1d8631b8c0da9d4e9d07. But I like the *iterator* idea.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java:
```diff
@@ -94,12 +96,58 @@ WithWarnings doMultipleWarningsArray(
+    } catch (UnsupportedMessageException | InvalidArrayIndexException e) {
+      throw CompilerDirectives.shouldNotReachHere(e);
+    }
+    var isLimitReached = curWarnsCnt >= maxWarns;
```
Why this is not using the `HashMapInsertAllNode`?

Akirathan replied:
Done in c2815e8a5d1064c2bbce1d8631b8c0da9d4e9d07

Commit:
HashMapInsertAll creates builder directly

Commit:
Transfer to interpreter before throwing an exception

JaroslavTulach commented:
@Akirathan, try to:
```diff
$ git diff
diff --git engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java
index 40b44fcb8f..23cdc559d1 100644
--- engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java
+++ engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java
@@ -1,5 +1,6 @@
 package org.enso.interpreter.runtime.warning;
 
+import com.oracle.truffle.api.CompilerAsserts;
 import com.oracle.truffle.api.CompilerDirectives;
 import com.oracle.truffle.api.dsl.Bind;
 import com.oracle.truffle.api.dsl.Cached;
@@ -68,6 +69,7 @@ public final class Warning implements EnsoObject {
 
   /** Slow version of {@link #fromMapToArray(EnsoHashMap, ArrayLikeLengthNode, ArrayLikeAtNode)}. */
   public static Warning[] fromMapToArray(EnsoHashMap map) {
+    CompilerAsserts.neverPartOfCompilation("Converting warngins to Warning[] is slow!");
     return fromMapToArray(map, ArrayLikeLengthNode.getUncached(), ArrayLikeAtNode.getUncached());
   }
 
```
that should fail every compilation that includes conversion of warnings to array (as it is slow). Then you'll know what to optimize.

Commit:
EnsoHashMap has iterator

Commit:
slow conversion from array to map calls CompilerAsserts.neverPartOfCompilation

JaroslavTulach committed:
javafmt

JaroslavTulach committed:
Avoid Arrays.copyOf in the IGV as vectorFromFunction always knows the right size

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java:
```diff
@@ -142,14 +142,31 @@ private Object get(int index, Node node) {
+  private Object toArray(boolean mustBeExact) {
     if (objectArray != null) {
-      return objectArray.length == size ? objectArray : Arrays.copyOf(objectArray, size);
+      return checkArraySize(mustBeExact, objectArray.length, size)
```
The graph we get is **too wide**:

![too wide IGV](https://github.com/user-attachments/assets/c34ab9ce-20b0-4643-80ad-a6b64c5c1339)

Ideal graph should _only contain branching when the user code does_ some `if` decisions. That's not the case right now.

9823d2ceae3472ec4227e3e06f876f3eb02d5e02 is first step towards that direction. In many cases (e.g.  `vectorFromAFunction`) we know the array is properly pre-allocated. We don't want `Arrays.copyOf` in the graph at all.

Akirathan commented:
One of the most suspicious thing about the `WarningBenchmark.diffWarningRandomElementsVecSum` benchmark is that when inspective IGV graphs, there is a one huge compilation unit that has inlined all the methods from the bench source https://github.com/enso-org/enso/blob/931baa42768febf7003455081c5354c22efe6b6e/engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/WarningBenchmarks.java#L81-L101. Including methods used to create the vector in the benchmark setup. Everything is inlined into `Value.execute`.

Trying to mitigate this with:
```diff
diff --git a/engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/WarningBenchmarks.java b/engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/WarningBenchmarks.java
index 24196a4dd..74771801d 100644
--- a/engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/WarningBenchmarks.java
+++ b/engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/WarningBenchmarks.java
@@ -7,6 +7,7 @@ import java.util.Objects;
 import java.util.Random;
 import java.util.concurrent.TimeUnit;
 import org.enso.common.MethodNames;
+import org.enso.common.MethodNames.Module;
 import org.enso.compiler.benchmarks.Utils;
 import org.graalvm.polyglot.Context;
 import org.graalvm.polyglot.Value;
@@ -33,6 +34,7 @@ public class WarningBenchmarks {
   private static final int INPUT_VEC_SIZE = 10_000;
   private static final int INPUT_DIFF_VEC_SIZE = 10_000;
   private Context ctx;
+  private Value assocType;
   private Value vecSumBench;
 
   private Value createVec;
@@ -108,26 +110,18 @@ public class WarningBenchmarks {
 
     var src = SrcUtil.source(benchmarkName, code.toString());
     Value module = ctx.eval(src);
-    vecSumBench =
-        Objects.requireNonNull(
-            module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "vec_sum_bench"));
-    createVec =
-        Objects.requireNonNull(
-            module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "create_vec"));
-    mapVecWithWarnings =
-        Objects.requireNonNull(
-            module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "map_vector_with_warnings"));
-    constElem =
-        Objects.requireNonNull(module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "elem"));
-    constElemWithWarning =
-        Objects.requireNonNull(
-            module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "elem_const_with_warning"));
+    assocType = module.invokeMember(MethodNames.Module.GET_ASSOCIATED_TYPE);
+    vecSumBench = Objects.requireNonNull(module.invokeMember(Module.GET_METHOD, assocType, "vec_sum_bench"));
+    createVec = Objects.requireNonNull(module.invokeMember(Module.GET_METHOD, assocType, "create_vec"));
+    mapVecWithWarnings = Objects.requireNonNull(module.invokeMember(Module.GET_METHOD, assocType, "map_vector_with_warnings"));
+    constElem = Objects.requireNonNull(module.invokeMember(Module.GET_METHOD, assocType, "elem"));
+    constElemWithWarning = Objects.requireNonNull(module.invokeMember(Module.GET_METHOD, assocType, "elem_const_with_warning"));
     noWarningsVec = createVec.execute(INPUT_VEC_SIZE, constElem);
     sameWarningVec = createVec.execute(INPUT_VEC_SIZE, constElemWithWarning);
     randomVec =
         Objects.requireNonNull(
             module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, randomIntVectorName));
-    randomElemsWithWarningsVec = mapVecWithWarnings.execute(randomVec);
+    randomElemsWithWarningsVec = mapVecWithWarnings.execute(assocType, randomVec);
   }
 
   @TearDown
@@ -155,7 +149,7 @@ public class WarningBenchmarks {
 
   @Benchmark
   public void diffWarningRandomElementsVecSum() {
-    Value res = vecSumBench.execute(randomElemsWithWarningsVec);
+    Value res = vecSumBench.execute(assocType, randomElemsWithWarningsVec);
     checkResult(res, randomVectorSum);
   }
```
fails.

JaroslavTulach committed:
CompilationFailureAction=Print is only useful when we enable logging

JaroslavTulach committed:
javafmt

JaroslavTulach commented:
With af7ac69 one can see that:
```bash
sbt:runtime-benchmarks> benchOnly diffWarning
```
bails out on **too deep inlining**:
```
[error] org.enso.interpreter.runtime.library.dispatch.TypeOfNodeGen.execute(TypeOfNodeGen.java:147)
[error] org.enso.interpreter.runtime.library.dispatch.TypeOfNodeGen.execute(TypeOfNodeGen.java:147)
[error] org.enso.interpreter.runtime.library.dispatch.TypeOfNodeGen.execute(TypeOfNodeGen.java:147)
[error] org.enso.interpreter.node.expression.builtin.meta.IsValueOfTypeNode.typeAndCheck(IsValueOfTypeNode.java:55)
[error] org.enso.interpreter.node.expression.builtin.meta.IsValueOfTypeNode$Typed.doType(IsValueOfTypeNode.java:123)
[error] org.enso.interpreter.node.expression.builtin.meta.IsValueOfTypeNodeGen$TypedNodeGen.execute(IsValueOfTypeNodeGen.java:302)
[error] org.enso.interpreter.node.expression.builtin.meta.IsValueOfTypeNodeGen.execute(IsValueOfTypeNodeGen.java:68)
[error] org.enso.interpreter.node.callable.argument.ReadArgumentCheckNode$TypeCheckNode.findDirectMatch(ReadArgumentCheckNode.java:367)
[error] org.enso.interpreter.node.callable.argument.ReadArgumentCheckNode$TypeCheckNode.doCheckNoConversionNeeded(ReadArgumentCheckNode.java:315)
[error] org.enso.interpreter.node.callable.argument.ReadArgumentCheckNodeFactory$TypeCheckNodeGen.executeCheckOrConversion(ReadArgumentCheckNodeFactory.java:112)
[error] org.enso.interpreter.node.callable.argument.ReadArgumentCheckNode.handleCheckOrConversion(ReadArgumentCheckNode.java:72)
[error] org.enso.interpreter.node.callable.argument.ReadArgumentNode.executeGeneric(ReadArgumentNode.java:72)
[error] org.enso.interpreter.node.scope.AssignmentNodeGen.executeGeneric(AssignmentNodeGen.java:51)
[error] org.enso.interpreter.node.BinaryOperatorNode.executeGeneric(BinaryOperatorNode.java:44)
[error] org.enso.interpreter.node.ClosureRootNode.execute(ClosureRootNode.java:85)
[error] com.oracle.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:745)
[error] com.oracle.truffle.runtime.OptimizedCallTarget.profiledPERoot(OptimizedCallTarget.java:669)
[error]         at jdk.internal.vm.compiler/org.graalvm.compiler.replacements.PEGraphDecoder.tooDeepInlining(PEGraphDecoder.java:1487)
```

JaroslavTulach committed:
Avoid recursive call in TypeOfNode.doWarning

JaroslavTulach committed:
Generate two source files: one for setup and other for benchmarking

JaroslavTulach committed:
Don't inline setup code

JaroslavTulach commented:
With f46b26b we are now slightly faster. With 6ebc286 the IGV shows only relevant pieces of the code. I can see one usage of `getUncached()` in the `diffWarnings` benchmark:

![getUncached](https://github.com/user-attachments/assets/57647a2e-3038-453c-b7af-d355abea484d)

It comes from `AddMethodGen` and is annotated with [TODO](https://github.com/enso-org/enso/pull/10555/files#diff-1ca826dc179449574ef71bd23d4c23451e5a9266bfc6caca9f098a20683256bbR633). Possibly the time has come to address that _TODO_ and things will be fast!

Akirathan commented:
@JaroslavTulach Thank you for the changes. I have updated my local copy of the branch, and the benchmark results are already 40% better! I will continue the investigation now.

Commit:
Warnings in builtin methods are collected via WarningsLibrary

Akirathan commented:
With 4005fe6d4c6f7ba35860682bf06c47a23884d766, the `WarningBenchmarks.diffWarningRandomElemensVecSum` score is now down to **14**. On develop, it is **176**. Hooray, we have the improvement that we were hunting for. @JaroslavTulach thank you for pointing out the obvious bottleneck.

Commit:
Remove unnecessary conversion to Warning[] array

Commit:
Remove unnecessary usages of ArrayRope

Commit:
Remove ArrayRope

Commit:
fmt

Commit:
Remove direct reference to generated node

Commit:
Rename AppendWarningNode.execute to executeAppend

Commit:
Cast long to int to not crash the interpreter

Commit:
fix typo in MethodProcessor

Commit:
Replace CompilerAsserts.neverPartOfCompilation with TruffleBoundary

JaroslavTulach commented on lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/MethodProcessor.java:
```diff
@@ -611,29 +628,27 @@ private boolean generateWarningsCheck(
             "    if ("
                 + arrayRead(argumentsArray, arg.getPosition())
-                + " instanceof WithWarnings) {");
+                + " instanceof WithWarnings withWarnings) {");
```
Recently I was claiming to @GregoryTravis that `instanceof WithWarnings` isn't the right way to check for warnings. Is or isn't it the right way in this case? Is or isn't it the right way in [Greg's case](https://github.com/enso-org/enso/pull/10725#discussion_r1701382652)?

Akirathan replied:
In this case, checking via `warnsLibrary.hasWarnings(x)` is not possible. The reason is that later, we fetch the value out of `withWarnings` via `withWarnings.getValue()`, which retains the old warnings. If you checked for warnings with `warnsLibrary.hasWarnings(x)`, you would need to fetch the value via `warnsLibrary.removeWarnings(x)`, and that does not retain old warnings. I have already tried to implement it that way, but some tests in `Warning_Spec` on `Warning.get_all wrap_errors=True` were failing.

Akirathan replied:
> Is or isn't it the right way in [Greg's case](https://github.com/enso-org/enso/pull/10725#discussion_r1701382652)?

Generally, I would prefer the check to be done via `WarningsLibrary`. This case in `MethodProcessor` is a bit special, though.

GregoryTravis replied:
On my branch, without these changes, the WarningsLibrary approach looks good:

https://github.com/enso-org/enso/pull/10725#discussion_r1705509913

Should I wait for this to be merged and benchmark my changes before merging them?

Akirathan replied:
@GregoryTravis I have just [scheduled benchmarks](https://github.com/enso-org/enso/pull/10555#issuecomment-2271731842) for this PR. All other tests are already successful. If benchmarks are OK, I will merge it. Wait, please before I compare the benchmarks, I will ping you then.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/EnsoHashMapBuilder.java:
```diff
@@ -82,6 +88,51 @@ public StorageEntry[] getEntries(int atGeneration, int size) {
     }
   }
 
+  public record Entry(Object key, Object value) {}
```
No public methods, classes, etc. in the `hash` package, please. I am not sure if this one is visible or not (because `EnshHashMapBuilder` isn't public), but still, the less `public` the better.

Akirathan replied:
All the public methods in `EnsoHashMap` and `EnsoHashMapBuilder` reverted to package-protected in these commits:
- 39f7262a7ed6772d542d043be3d9ec63f7ec0353
- 2d7a8b3c571e6545d6ed8cdb065179f897019d70
- 0ecab16a50462fecb3ae1cfae0fcaca0c9f53ac3

Commit:
tests in interpreter-dsl-test enter polyglot context

Commit:
Methods in EnsoHashMapBuilder are package-protected

Commit:
TailCallException stores warnings as EnsoHashMap

Commit:
Make EnsoHashMap.getCachedVectorRepresentation methods package-protected

Commit:
EnsoHashMap.getHashSize is package-protected.

Hash size is fetched via HashMapSizeNode.

Commit:
fmt

Akirathan commented:
- Engine benchmarks scheduled at: https://github.com/enso-org/enso/actions/runs/10270716008
- Stdlib benchmarks scheduled at: https://github.com/enso-org/enso/actions/runs/10270723661

<blockquote><img src="https://opengraph.githubassets.com/d385509ef36d0cb1d4a02c885f3e183af15253f31f167e8b7960bb30dd85ca3c/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@a5878b2</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>
<blockquote><img src="https://opengraph.githubassets.com/d385509ef36d0cb1d4a02c885f3e183af15253f31f167e8b7960bb30dd85ca3c/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Standard Libraries · enso-org/enso@a5878b2</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeConversionNode.java:
```diff
@@ -157,9 +159,13 @@ Object doWarning(
+      @CachedLibrary(limit = "3") InteropLibrary interop) {
     arguments[thatArgumentPosition] = that.getValue();
-    ArrayRope<Warning> warnings = that.getReassignedWarningsAsRope(this, false);
+    Warning[] warnings = that.getWarningsArray(false, warnsLib, mapInsertAllNode, interop);
```
This should probably also operate on `EnsoHashMap` and not on `Warning[]`...

Akirathan replied:
Right. Done in a88ae7489b6778b02ac5e66266620a982b53b113.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeMethodNode.java:
```diff
@@ -129,9 +131,13 @@ Object doWarning(
+      @CachedLibrary(limit = "3") InteropLibrary interop) {
     arguments[thisArgumentPosition] = self.getValue();
-    ArrayRope<Warning> warnings = self.getReassignedWarningsAsRope(this, false);
+    Warning[] warnings = self.getWarningsArray(false, warnsLib, mapInsertAllNode, interop);
```
Another place for `EnsoHashMap` I think.

Akirathan replied:
Right. Migrated to `EnsoHashMap` in a88ae7489b6778b02ac5e66266620a982b53b113

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/EnsoHashMapBuilder.java:
```diff
@@ -82,12 +88,57 @@ public StorageEntry[] getEntries(int atGeneration, int size) {
+    return new EntriesIterator(atGeneration);
+  }
+
+  class EntriesIterator implements Iterator<Entry> {
```
Thanks for removing any signs of `public` in this and `EnsoHashMap` files. Possibly you can also make this class `final`.

Akirathan replied:
Made final in a44cf5af6aab98c5d65ea7269d798f86acff64fd

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/HashMapInsertAllNode.java:
```diff
@@ -0,0 +1,73 @@
+      VirtualFrame frame, Object self, Object container, int maxItems);
+
+  @Specialization
+  EnsoHashMap doEnsoHashMaps(
```
Only a single specialization and no fallback? Have you tried to change signature of `executeInsertAll` to accept only `EnsoHashMap`s?

Akirathan replied:
Good observation. Signature changed in ac6dfb42c6c7f983dfac0513927c4841d115ddcd

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/HashMapInsertAllNode.java:
```diff
@@ -0,0 +1,73 @@
+    if (otherSize == 0) {
+      return self;
+    }
+    var mapBuilder = EnsoHashMapBuilder.createWithCapacity(selfSize + otherSize);
```
This seems to create an empty builder and then it inserts entries from `self` and `other` into it? Why don't we just **insert** entries from `other` into/onto `self`? Or rather select the smaller map and insert entries from the smaller map onto the bigger one!

Akirathan replied:
I have already tried that before. The problem is *rehashing*. If you don't create a sufficiently big hash map beforehand, you will inevitably have to rehash. With this approach, you don't have to rehash.

JaroslavTulach reviewed (approved):
This is a huge change and it'd be better to get it finally in. These are the benchmarks results I see:
```
sbt:runtime-benchmarks> benchOnly Warning
[info] Benchmark                                          Mode  Cnt   Score   Error  Units
[info] WarningBenchmarks.diffWarningRandomElementsVecSum  avgt    3  12.885 ± 3.815  ms/op
[info] WarningBenchmarks.noWarningsVecSum                 avgt    3   0.006 ± 0.001  ms/op
[info] WarningBenchmarks.randomElementsVecSum             avgt    3   0.034 ± 0.014  ms/op
[info] WarningBenchmarks.sameWarningVecSum                avgt    3   1.842 ± 0.306  ms/op
```
It is far from ideal, but if we are faster than current `develop` and CI is green, let's integrate and continue this endeavor in another PR.

Akirathan commented:
Comparing the [benchmark results](https://github.com/enso-org/enso/pull/10555#issuecomment-2271731842) with develop:
- In engine benchmarks, I can see that all `WarningBenchmarks` results are better than on develop. More specifically:
  - `WarningBenchmarks.diffWarningRandomElementsVecSum`: 137 vs 27
  - `WarningBenchmarks.noWarningsVecSum`: 0.16 vs 0.0031
  - `WarningBenchmarks.randomElementsVecSum`: 0.206 vs 0.022
  - `WarningBenchmarks.sameWarningVecSum`: 8.311 vs 1.757
- stdlib benchmarks are without any regressions.

Conclussion: Benchmarks look very good.

Commit:
EntriesIterator is final

Commit:
Use EnsoHashMap for warnings instead of an array.

Commit:
Change signature of HashMapInsertAllNode to accept only EnsoHashMaps

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/HashMapInsertAllNode.java:
```diff
@@ -0,0 +1,73 @@
+    var mapBuilder = EnsoHashMapBuilder.createWithCapacity(selfSize + otherSize);
+
+    var selfMapBuilder = self.getMapBuilder(frame, true, hashCodeNode, equalsNode);
+    var selfEntriesIt = selfMapBuilder.getEntriesIterator(selfMapBuilder.generation());
```
Using `selfMapBuilder.generation()` isn't correct. Fixed in https://github.com/enso-org/enso/pull/12201/files#r1946171034
